Add "Introduction to Functional Programming and the Structure of Programming Languages using OCaml" to books section (#1744)
ocaml/ocaml.org: data/books/introduction-to-functional-programming-and-the-structure-of-programming-languages-using-ocaml.md

title: "Introduction to Functional Programming and the Structure of Programming Languages using OCaml"
slug: "intro-to-fp-and-structure-of-pl"
description: >
  This text teaches functional programming and the structure of programming languages to beginning students.
  It is written for the Programming 1 course for computer science students at Saarland University.
  We assume that incoming students are familiar with mathematical thinking, but we do not assume programming experience.
authors:
  - Gert Smolka
published: 2022-04-05
cover: books/intro-to-fp-and-structure-of-pl.jpg
language:
  - english
links:
  - description: GitHub repository
    uri: https://github.com/uds-psl/Prog
  - description: PDF download
    uri: https://www.ps.uni-saarland.de/~smolka/drafts/prog2021.pdf
rating: 5
featured: false
difficulty: beginner
pricing: free
---

This text teaches functional programming and the structure of programming languages to beginning students.
It is written for the Programming 1 course for computer science students at Saarland University.
We assume that incoming students are familiar with mathematical thinking,
but we do not assume programming experience. The course is designed

> to take about one third of the study time of the first semester.

Reviewer: ```suggestion
to take about one third of the first semester's study time.
```

As it comes to functional programming, we cover higher-order recursive functions, polymorphic typing, and constructor types for lists,

> trees, and abstract syntax. We emphasize the role of correctness statements and practice inductive correctness proofs. We also cover asymptotic running time considering binary search (logarithmic), insertion sort

Reviewer: ```suggestion
trees, and abstract syntax. We emphasise the role of correctness statements and practice inductive correctness proofs. We also cover asymptotic running time considering binary search (logarithmic), insertion sort
```

(quadratic), merge sort (linearithmic), and other algorithms.

As it comes to the structure of programming languages, we study

> the different layers of syntax and semantics at the example of the idealized functional programming language Mini-OCaml. We describe the
> syntactic layers with grammars and the semantic layers with inference
> rules. Based on these formal descriptions, we program recursive descent
> parsers, type checkers and evaluators.

Reviewer: ```suggestion
the different layers of syntax and semantics at the example of the idealised functional programming language Mini-OCaml. We describe the
syntactic layers with grammar and the semantic layers with inference
rules. Based on these formal descriptions, we program recursive descent
parsers, type checkers and evaluators.
```
